strategoxt/stratego-libraries/lib/spec/term/string.str

```diff
@@ -22,12 +22,74 @@ imports
 strategies
 
   /**
+   * Scopes local-newname and local-new, which work by dynamic rule.
+   */
+  scope-local-new(s) = {| LocalNew, LocalNewAlpha : rules(LocalNewAlpha := 0);  s |}
+
+  /**
+   * Generates a locally unique new string with user-defined prefix
+   * the form 'sd', where s is a string and d is a positive integer.
+   * The produced string is *only* unique within the scope of
+   * scope-local-new, and only based on previous calls to
+   * local-newname/local-new. If you constructed a similar name by
+   * hand it will not be avoided.
+   *
+   * The prefix comes in front of an increasing number, which is
+   * separately maintained and started at 0 for each prefix.
+   *
+   * Examples (results are shown in comment)
+   *   scope-local-new(
+   *     <local-newname; debug> "a"        // produces "a0"
+   *   ; <local-newname; debug> "a"        // produces "a1"
+   *   ; <local-newname; debug> "a"        // produces "a2"
+   *   ; <local-newname; debug> "b"        // produces "b0"
+   *   ; <local-newname; debug> "b"        // produces "b1"
+   *   ; <local-newname; debug> "b_2"      // produces "b_20"
+   *   ; <local-newname; debug> "b_2"      // produces "b_21"
+   *   ; <local-newname; debug> "b_a"      // produces "b_a0"
+   *   ; <local-newname; debug> "b_a"      // produces "b_a1"
+   *   ; <local-newname; debug> "a_"       // produces "a_0"
+   *   ; <local-newname; debug> "a_1"      // produces "a_10"
+   *   )
+   */
+  local-newname: prefix -> $[[prefix][number]]
+  where number := <LocalNew <+ !0> prefix
+      ; number' := <int-add(|1)> number
+      ; rules(LocalNew: prefix -> number')
+
+  /**
+   * Generates a locally unique new string of the form 'c_d',
+   * where c is one char and d is a positive integer.
+   * The produced string is *only* unique within the scope of
+   * scope-local-new, and only based on previous calls to
+   * local-newname/local-new. If you constructed a similar name by
+   * hand it will not be avoided.
+   */
+  local-new =
```

Comment:
Why not write this as a rule like `local-newname`? I can't understand what is going on in this strategy.

Reply:
Here's a little example program:
```stratego
module test-local-new

imports

  libstratego-lib

strategies

main = scope-local-new(repeat(local-new;debug|30))
```

And this is the output:

```stratego
a_0
b_0
c_0
d_0
e_0
f_0
g_0
h_0
i_0
j_0
k_0
l_0
m_0
n_0
o_0
p_0
q_0
r_0
s_0
t_0
u_0
v_0
w_0
x_0
y_0
z_0
a_1
b_1
c_1
d_1
```

Reply:
I've added some line comments to explain the implementation